File: sdks/official/fraiseql-typescript/tests/rest-annotations.test.ts
import { describe, it, expect, beforeEach } from "vitest";
import { SchemaRegistry } from "../src/registry";
import { registerQuery, registerMutation } from "../src/decorators";

describe("REST annotations", () => {
  beforeEach(() => {
    SchemaRegistry.clear();
  });

  describe("queries", () => {
    it("should include rest block when restPath is set", () => {
      registerQuery("users", "User", true, false, [], "Get users", {
        sqlSource: "v_user",
        restPath: "/api/users",
        restMethod: "GET",
      });
      const schema = SchemaRegistry.getSchema();
      expect(schema.queries[0].rest).toEqual({ path: "/api/users", method: "GET" });
    });

    it("should default method to GET for queries", () => {
      registerQuery("users", "User", true, false, [], "Get users", {
        sqlSource: "v_user",
        restPath: "/api/users",
      });
      const schema = SchemaRegistry.getSchema();
      expect(schema.queries[0].rest).toEqual({ path: "/api/users", method: "GET" });
    });

    it("should omit rest block when restPath is not set", () => {
      registerQuery("users", "User", true, false, [], "Get users", {
        sqlSource: "v_user",
      });
      const schema = SchemaRegistry.getSchema();
      expect(schema.queries[0].rest).toBeUndefined();
    });

    it("should uppercase the method", () => {
      registerQuery("users", "User", true, false, [], "Get users", {
        sqlSource: "v_user",
        restPath: "/api/users",
        restMethod: "post",
      });
      const schema = SchemaRegistry.getSchema();
      expect(schema.queries[0].rest?.method).toBe("POST");
    });
  });

  describe("mutations", () => {
    it("should include rest block when restPath is set", () => {
      registerMutation("createUser", "User", false, false, [], "Create user", {
        sqlSource: "fn_create_user",
        operation: "CREATE",
        restPath: "/api/users",
        restMethod: "POST",
      });
      const schema = SchemaRegistry.getSchema();
      expect(schema.mutations[0].rest).toEqual({ path: "/api/users", method: "POST" });
    });

    it("should default method to POST for mutations", () => {
      registerMutation("createUser", "User", false, false, [], "Create user", {
        sqlSource: "fn_create_user",
        operation: "CREATE",
        restPath: "/api/users",
      });
      const schema = SchemaRegistry.getSchema();
      expect(schema.mutations[0].rest?.method).toBe("POST");
    });

    it("should support DELETE method", () => {
      registerMutation("deleteUser", "User", false, false, [], "Delete user", {
        sqlSource: "fn_delete_user",
        operation: "DELETE",
        restPath: "/api/users/{id}",
        restMethod: "DELETE",
      });
      const schema = SchemaRegistry.getSchema();
      expect(schema.mutations[0].rest).toEqual({ path: "/api/users/{id}", method: "DELETE" });
    });

    it("should omit rest block when restPath is not set", () => {
      registerMutation("createUser", "User", false, false, [], "Create user", {
        sqlSource: "fn_create_user",
        operation: "CREATE",
      });
      const schema = SchemaRegistry.getSchema();
      expect(schema.mutations[0].rest).toBeUndefined();
    });
  });

  describe("validation", () => {
    it("should throw when restMethod is set without restPath", () => {
      expect(() => {
        registerQuery("users", "User", true, false, [], "Get users", {
          sqlSource: "v_user",
          restMethod: "GET",
        });
      }).toThrow("restMethod requires restPath to be set");
    });

    it("should throw for invalid REST method", () => {
      expect(() => {
        registerQuery("users", "User", true, false, [], "Get users", {
          sqlSource: "v_user",
          restPath: "/api/users",
          restMethod: "INVALID",
        });
      }).toThrow("Invalid REST method");
    });
  });
});

File: sdks/official/fraiseql-go/fraiseql/rest_annotations_test.go
package fraiseql

import (
	"encoding/json"
	"testing"
)

func TestQueryRestAnnotation(t *testing.T) {
	Reset()
	defer Reset()
	err := NewQuery("users").
		ReturnType("User").
		ReturnsArray(true).
		SqlSource("v_user").
		RestPath("/api/users").
		RestMethod("GET").
		Register()
	if err != nil {
		t.Fatalf("Register failed: %v", err)
	}

	schema := GetSchema()
	data, _ := json.Marshal(schema)
	var result map[string]interface{}
	json.Unmarshal(data, &result)

	queries := result["queries"].([]interface{})
	query := queries[0].(map[string]interface{})
	rest := query["rest"].(map[string]interface{})
	if rest["path"] != "/api/users" {
		t.Errorf("expected path /api/users, got %v", rest["path"])
	}
	if rest["method"] != "GET" {
		t.Errorf("expected method GET, got %v", rest["method"])
	}
}

func TestQueryRestDefaultMethodGET(t *testing.T) {
	Reset()
	defer Reset()
	err := NewQuery("users").
		ReturnType("User").
		ReturnsArray(true).
		SqlSource("v_user").
		RestPath("/api/users").
		Register()
	if err != nil {
		t.Fatalf("Register failed: %v", err)
	}

	schema := GetSchema()
	data, _ := json.Marshal(schema)
	var result map[string]interface{}
	json.Unmarshal(data, &result)

	queries := result["queries"].([]interface{})
	query := queries[0].(map[string]interface{})
	rest := query["rest"].(map[string]interface{})
	if rest["method"] != "GET" {
		t.Errorf("expected default method GET, got %v", rest["method"])
	}
}

func TestQueryWithoutRestOmitsBlock(t *testing.T) {
	Reset()
	defer Reset()
	err := NewQuery("users").
		ReturnType("User").
		ReturnsArray(true).
		SqlSource("v_user").
		Register()
	if err != nil {
		t.Fatalf("Register failed: %v", err)
	}

	schema := GetSchema()
	data, _ := json.Marshal(schema)
	var result map[string]interface{}
	json.Unmarshal(data, &result)

	queries := result["queries"].([]interface{})
	query := queries[0].(map[string]interface{})
	if _, ok := query["rest"]; ok {
		t.Error("expected no rest block when restPath not set")
	}
}

func TestMutationRestAnnotation(t *testing.T) {
	Reset()
	defer Reset()
	err := NewMutation("createUser").
		ReturnType("User").
		SqlSource("fn_create_user").
		Operation("insert").
		RestPath("/api/users").
		RestMethod("POST").
		Register()
	if err != nil {
		t.Fatalf("Register failed: %v", err)
	}

	schema := GetSchema()
	data, _ := json.Marshal(schema)
	var result map[string]interface{}
	json.Unmarshal(data, &result)

	mutations := result["mutations"].([]interface{})
	mutation := mutations[0].(map[string]interface{})
	rest := mutation["rest"].(map[string]interface{})
	if rest["path"] != "/api/users" {
		t.Errorf("expected path /api/users, got %v", rest["path"])
	}
	if rest["method"] != "POST" {
		t.Errorf("expected method POST, got %v", rest["method"])
	}
}

func TestMutationRestDefaultMethodPOST(t *testing.T) {
	Reset()
	defer Reset()
	err := NewMutation("createUser").
		ReturnType("User").
		SqlSource("fn_create_user").
		Operation("insert").
		RestPath("/api/users").
		Register()
	if err != nil {
		t.Fatalf("Register failed: %v", err)
	}

	schema := GetSchema()
	data, _ := json.Marshal(schema)
	var result map[string]interface{}
	json.Unmarshal(data, &result)

	mutations := result["mutations"].([]interface{})
	mutation := mutations[0].(map[string]interface{})
	rest := mutation["rest"].(map[string]interface{})
	if rest["method"] != "POST" {
		t.Errorf("expected default method POST, got %v", rest["method"])
	}
}

func TestMutationRestDeleteMethod(t *testing.T) {
	Reset()
	defer Reset()
	err := NewMutation("deleteUser").
		ReturnType("User").
		SqlSource("fn_delete_user").
		Operation("delete").
		RestPath("/api/users/{id}").
		RestMethod("DELETE").
		Register()
	if err != nil {
		t.Fatalf("Register failed: %v", err)
	}

	schema := GetSchema()
	data, _ := json.Marshal(schema)
	var result map[string]interface{}
	json.Unmarshal(data, &result)

	mutations := result["mutations"].([]interface{})
	mutation := mutations[0].(map[string]interface{})
	rest := mutation["rest"].(map[string]interface{})
	if rest["path"] != "/api/users/{id}" {
		t.Errorf("expected path /api/users/{id}, got %v", rest["path"])
	}
	if rest["method"] != "DELETE" {
		t.Errorf("expected method DELETE, got %v", rest["method"])
	}
}

func TestRestMethodCaseInsensitive(t *testing.T) {
	Reset()
	defer Reset()
	err := NewQuery("users").
		ReturnType("User").
		ReturnsArray(true).
		SqlSource("v_user").
		RestPath("/api/users").
		RestMethod("get").
		Register()
	if err != nil {
		t.Fatalf("Register failed: %v", err)
	}

	schema := GetSchema()
	data, _ := json.Marshal(schema)
	var result map[string]interface{}
	json.Unmarshal(data, &result)

	queries := result["queries"].([]interface{})
	query := queries[0].(map[string]interface{})
	rest := query["rest"].(map[string]interface{})
	if rest["method"] != "GET" {
		t.Errorf("expected uppercase method GET, got %v", rest["method"])
	}
}
